pkg/apps/deployment/status_test.go

```diff
@@ -0,0 +1,223 @@
+	pod.Name = "test"
+	mutateFn(pod)
+	return pod
+}
+
+type fakePodLister struct {
+	pods []*corev1.Pod
+}
+
+func (f *fakePodLister) List(selector labels.Selector) ([]*corev1.Pod, error) {
+	panic("implement me")
+}
+
+func (f *fakePodLister) Pods(namespace string) v1.PodNamespaceLister {
+	return &fakePodNamespacer{
+		pods: f.pods,
+	}
+}
+
+type fakePodNamespacer struct {
+	pods []*corev1.Pod
+}
+
+func (f *fakePodNamespacer) List(selector labels.Selector) ([]*corev1.Pod, error) {
+	return f.pods, nil
+}
+
+func (f *fakePodNamespacer) Get(name string) (*corev1.Pod, error) {
+	panic("implement me")
+}
+
+func TestPodContainersStatus(t *testing.T) {
+	var (
+		defaultLabels = map[string]string{"app": "test"}
+	)
+	tests := []struct {
+		name             string
+		pods             []*corev1.Pod
+		deployment       *appsv1.Deployment
+		expectedMessages []string
+		expectedError    error
+	}{
+		{
+			name: "ready with one restart",
+			pods: append([]*corev1.Pod{}, makePod(func(pod *corev1.Pod) {
+				pod.Labels = defaultLabels
+				pod.Status.Phase = corev1.PodRunning
+				pod.Status.ContainerStatuses = append(pod.Status.ContainerStatuses, corev1.ContainerStatus{
+					Name:         "test",
+					RestartCount: 1,
+					Ready:        true,
+				})
+			})),
+			deployment:       makeDeployment("foo", defaultLabels),
+			expectedMessages: []string{},
+		},
+		{
+			name: "not ready with one restart",
+			pods: append([]*corev1.Pod{}, makePod(func(pod *corev1.Pod) {
+				pod.Labels = defaultLabels
+				pod.Status.Phase = corev1.PodRunning
+				pod.Status.ContainerStatuses = append(pod.Status.ContainerStatuses, corev1.ContainerStatus{
+					Name:         "test",
+					RestartCount: 1,
+				})
+			})),
+			deployment:       makeDeployment("foo", defaultLabels),
+			expectedMessages: []string{"1 container is not ready in pod test"},
+		},
+		{
+			name: "not ready with two restarts",
+			pods: append([]*corev1.Pod{}, makePod(func(pod *corev1.Pod) {
+				pod.Labels = defaultLabels
+				pod.Status.Phase = corev1.PodRunning
+				pod.Status.ContainerStatuses = append(pod.Status.ContainerStatuses, corev1.ContainerStatus{
+					Name:         "test",
+					RestartCount: 2,
+				})
+			})),
+			deployment:       makeDeployment("foo", defaultLabels),
+			expectedMessages: []string{"1 container is crashlooping in pod test"},
+		},
+		{
+			name: "not ready and stucked in waiting",
+			pods: append([]*corev1.Pod{}, makePod(func(pod *corev1.Pod) {
+				pod.Labels = defaultLabels
+				pod.Status.Phase = corev1.PodPending
+				pod.Status.ContainerStatuses = append(pod.Status.ContainerStatuses, corev1.ContainerStatus{
+					Name: "test",
+					State: corev1.ContainerState{
+						Waiting: &corev1.ContainerStateWaiting{
+							Reason:  "IsSlow",
+							Message: "Slow container",
+						},
+					},
+				})
+			})),
+			deployment:       makeDeployment("foo", defaultLabels),
+			expectedMessages: []string{"1 container is waiting in pending pod test"},
```

Comment:
nit: drop "1"

Reply:
i like to know how many :-)

Reply:
`1 container is waiting` does not contradict that there is another one. Remember I am mathematician.

Reply:
can't argue with that. changed.